### PR TITLE
TDL-18660 Add retry mechanism and error logging

### PR DESCRIPTION
**Pull request by karanpanchal-crest**
# Description of change
 - Add error logging message for all exception
 - Add unit test to verify that proper message is logged when exception raised

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_snapchat_ads/client.py [hunk 1]
```diff
@@ -58,25 +56,63 @@ class SnapchatForbiddenError(SnapchatError):
...
+        "raise_exception": SnapchatNotFoundError, 
+        "message":"The resource you have specified cannot be found."
+    },
+    405: {
+        "raise_exception": SnapchatMethodNotAllowedError, 
+        "message":"The provided HTTP method is not supported by the URL."
+    },
+    406: {"raise_exception": SnapchatNotAcceptableError, 
+          "message":"You requested a format that isn’t json."
+    },
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/client.py:90]:
```suggestion
    406: {
          "raise_exception": SnapchatNotAcceptableError, 
          "message":"You requested a format that isn’t json."
    },
```

File: tap_snapchat_ads/client.py [hunk 2]
```diff
@@ -58,25 +56,63 @@ class SnapchatForbiddenError(SnapchatError):
...
+        "message": "You are requesting to many requests."
+    },
+    500: {
+        "raise_exception": SnapchatInternalServiceError,
+        "message": "Internal Server Error."
+    },
+    503: {
+        "raise_exception": SnapchatServiceUnavailableError,
+        "message": "Service Unavailable."
+    }}
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/client.py:110]:
```suggestion
    }
}
```

File: tap_snapchat_ads/client.py [hunk 3]
```diff
@@ -58,25 +56,63 @@ class SnapchatForbiddenError(SnapchatError):
...
+        "raise_exception": SnapchatTeapotError, 
+        "message":"The server refuses to brew coffee because it is, permanently, a teapot."
+    },
+    429: {
+        "raise_exception": Server429Error,
+        "message": "You are requesting to many requests."
+    },
+    500: {
+        "raise_exception": SnapchatInternalServiceError,
+        "message": "Internal Server Error."
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/client.py:105]:
```suggestion
        "message": "An error has occurred at Snapchat's end."
```

File: tap_snapchat_ads/client.py [hunk 4]
```diff
@@ -58,25 +56,63 @@ class SnapchatForbiddenError(SnapchatError):
...
+    },
+    503: {
+        "raise_exception": SnapchatServiceUnavailableError,
+        "message": "Service Unavailable."
```
Comment (savan-chovatiya) [inline tap_snapchat_ads/client.py:109]:
```suggestion
        "message": "API service is currently unavailable."
```

**Review by savan-chovatiya — CHANGES_REQUESTED** [review]

File: tap_snapchat_ads/client.py (hunk 1 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/client.py]: Updated code

File: tap_snapchat_ads/client.py (hunk 2 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/client.py]: Updated code

File: tap_snapchat_ads/client.py (hunk 3 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/client.py]: Updated message as suggested

File: tap_snapchat_ads/client.py (hunk 4 above)
Comment (karanpanchal-crest) [reply tap_snapchat_ads/client.py]: Updated message as suggested

**Review by savan-chovatiya — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Review by RushiT0122 — APPROVED** [review]

**Pull request merged**